src/pgrepr/src/value/apd.rs

```diff
@@ -0,0 +1,242 @@
+            cx.div_integer(&mut w, &TO_SQL_BASER);
+            d = w;
+        }
+        d_i -= leading_zero_units;
+
+        let units = u16::try_from(UNITS_LEN - d_i).unwrap();
+        let weight = i16::try_from(units - fract_units).unwrap() - 1;
+
+        out.put_u16(units);
+        out.put_i16(weight);
+        // sign
+        out.put_u16(if is_nan {
+            0xC000
+        } else if is_neg {
+            0x4000
+        } else {
+            0
+        });
+        out.put_u16(scale);
+        if !is_nan {
+            for digit in digits[d_i..].iter() {
+                out.put_u16(*digit);
+            }
+        }
+
+        Ok(IsNull::No)
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        matches!(*ty, Type::NUMERIC)
+    }
+
+    to_sql_checked!();
+}
+
+impl<'a> FromSql<'a> for Apd {
+    fn from_sql(_: &Type, mut raw: &'a [u8]) -> Result<Apd, Box<dyn Error + Sync + Send>> {
+        let units = usize::from(raw.read_u16::<NetworkEndian>()?);
+        let weight = raw.read_i16::<NetworkEndian>()?;
+        let sign = raw.read_u16::<NetworkEndian>()?;
+        let in_scale = raw.read_i16::<NetworkEndian>()?;
+        let mut digits = vec![];
+        for _ in 0..units {
+            digits.push(raw.read_u16::<NetworkEndian>()?)
+        }
+
+        // We need wider context because decoding values can require >39 digits
+        // of precision given how alignment works.
+        let mut cx = apd::cx_agg();
+        let mut d = ApdAgg::zero();
+
+        for digit in digits[..units].iter() {
+            cx.scaleb(&mut d, &FROM_SQL_SCALER);
+            let n = AdtApd::from(u32::from(*digit));
+            cx.add(&mut d, &n);
+        }
+
+        match sign {
+            0 => (),
+            0x4000 => cx.neg(&mut d),
+            0xC000 => return Ok(Apd(OrderedDecimal(AdtApd::nan()))),
+            _ => return Err("bad sign in numeric".into()),
+        }
+
+        let mut scale = (units as i16 - weight - 1) * 4;
+
+        // Adjust scales
+        if scale < 0 {
+            // Multiply by 10^scale
+            cx.scaleb(&mut d, &AdtApd::from(-i32::from(scale)));
+            scale = 0;
+        } else if scale > in_scale {
+            // Divide by 10^(difference in scale and in_scale)
+            cx.scaleb(&mut d, &AdtApd::from(-i32::from(scale - in_scale)));
+            scale = in_scale;
+        }
+
+        cx.scaleb(&mut d, &AdtApd::from(-i32::from(scale)));
+        cx.reduce(&mut d);
+
+        let mut cx = cx_datum();
+        let d_datum = cx.to_width(d);
+
+        // Reducing before taking to datum width lets us check for any status
+        // for errors.
+        if d.is_infinite() || cx.status().any() {
+            return Err(format!("Unable to take bytes to APD value; rendered {}", d).into());
+        }
+        Ok(Apd(OrderedDecimal(d_datum)))
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        matches!(*ty, Type::NUMERIC)
+    }
+}
+
+#[test]
+fn test_to_from_sql_roundtrip() {
```

Comment:
Thank you for the tests . Can you consider some more scenarios -- I did not push this patch directly as I am not sure which branch you prefer to work in:


```
diff --git a/src/pgrepr/src/value/apd.rs b/src/pgrepr/src/value/apd.rs
index 6a2e4b70c..b9603a1b9 100644
--- a/src/pgrepr/src/value/apd.rs
+++ b/src/pgrepr/src/value/apd.rs
@@ -224,6 +224,8 @@ fn test_to_from_sql_roundtrip() {
         let d_from_sql = Apd::from_sql(&Type::NUMERIC, &out).unwrap();
         assert_eq!(r.0, d_from_sql.0);
     }
+    inner("0");
+    inner("-0");
     inner("0.1");
     inner("0.0");
     inner("0.00");
@@ -231,12 +233,50 @@ fn test_to_from_sql_roundtrip() {
     inner("0.0000");
     inner("0.00000");
     inner("123456789.012346789");
+    inner("000000000000000000000000000000000000001");
+    inner("000000000000000000000000000000000000000");
+    inner("999999999999999999999999999999999999999");
     inner("123456789012345678901234567890123456789");
+    inner("-123456789012345678901234567890123456789");
     inner(".123456789012345678901234567890123456789");
+    inner(".000000000000000000000000000000000000001");
+    inner(".000000000000000000000000000000000000000");
+    inner(".999999999999999999999999999999999999999");
+    inner("-0.123456789012345678901234567890123456789");
     inner("1e25");
+    inner("-1e25");
     inner("9.876e-25");
     inner("-9.876e-25");
     inner("98760000");
     inner(".00009876");
+    inner("-.00009876");
     inner("NaN");
 }
```

Reply:
Added these locally; will include them in the next push for this PR